sklearn/metrics/base.py

```diff
@@ -131,3 +132,69 @@ def _average_binary_score(binary_metric, y_true, y_score, average,
         return np.average(score, weights=average_weight)
     else:
         return score
+
+
+def _average_multiclass_ovo_score(binary_metric, y_true, y_score, average):
+    """Uses the binary metric for one-vs-one multiclass classification,
+    where the score is computed according to the Hand & Till (2001) algorithm.
+
+    Parameters
+    ----------
+    y_true : array, shape = [n_samples]
+        True multiclass labels.
+        Assumes labels have been recoded to 0 to n_classes.
+
+    y_score : array, shape = [n_samples, n_classes]
+        Target scores corresponding to probability estimates of a sample
+        belonging to a particular class
+
+    average : 'macro' or 'weighted', default='macro'
+        ``'macro'``:
+            Calculate metrics for each label, and find their unweighted
+            mean. This does not take label imbalance into account. Classes
+            are assumed to be uniformly distributed.
+        ``'weighted'``:
+            Calculate metrics for each label, taking into account the a priori
+            distribution of the classes.
+
+    binary_metric : callable, the binary metric function to use.
+        Accepts the following as input
+            y_true_target : array, shape = [n_samples_target]
+                Some sub-array of y_true for a pair of classes designated
+                positive and negative in the one-vs-one scheme.
+            y_score_target : array, shape = [n_samples_target]
+                Scores corresponding to the probability estimates
+                of a sample belonging to the designated positive class label
+
+    Returns
+    -------
+    score : float
+        Average the sum of pairwise binary metric scores
+    """
+    n_labels = len(np.unique(y_true))
+    pos_and_neg_prevalence = []
+    label_scores = []
+    for pos, neg in itertools.combinations(range(n_labels), 2):
+        pos_ix = y_true == pos
+        ix = np.logical_or(pos_ix, y_true == neg)
+
+        pos_and_neg_prevalence.append(float(np.sum(ix)) / len(y_true))
+
+        y_score_filtered = y_score[ix]
+
+        class_a = pos_ix[ix]
+        class_b = np.logical_not(class_a)
+
+        score_class_a = binary_metric(
+                class_a, y_score_filtered[:, pos])
+        score_class_b = binary_metric(
+                class_b, y_score_filtered[:, neg])
+        binary_avg_score = (score_class_a + score_class_b) / 2.
```

Comment:
This is based on [pg. 177](http://download.springer.com/static/pdf/398/art%253A10.1023%252FA%253A1010920819831.pdf?originUrl=http%3A%2F%2Flink.springer.com%2Farticle%2F10.1023%2FA%3A1010920819831&token2=exp=1489175909~acl=%2Fstatic%2Fpdf%2F398%2Fart%25253A10.1023%25252FA%25253A1010920819831.pdf%3ForiginUrl%3Dhttp%253A%252F%252Flink.springer.com%252Farticle%252F10.1023%252FA%253A1010920819831*~hmac=3e1dfe09d9d14b8a64343eaa27d2b06bd5e7d9cfa7ebb66da983de5ca5f41d7d) of Hand-Till 2001, 4 lines above equation (7).